Tests/RevenueCatUITests/Templates/ExternalPurchaseAndRestoreTests.swift

```diff
@@ -0,0 +1,393 @@
+            offering: Self.offering.withLocalImages,
+            customerInfo: TestData.customerInfo,
+            introEligibility: .producing(eligibility: .eligible),
+            purchaseHandler: purchasHandler
+        )
+        .onPurchaseStarted { _ in
+            callbackOrder.append("onPurchaseStarted")
+        }
+        .onPurchaseCompleted { _ in
+            callbackOrder.append("onPurchaseCompleted")
+        }
+        .onPurchaseCancelled {
+            callbackOrder.append("onPurchaseCancelled")
+        }
+        .onPurchaseFailure { error in
+            callbackOrder.append("onPurchaseFailure")
+            passedError = error
+        }
+        .onRestoreStarted({
+            callbackOrder.append("onRestoreStarted")
+        })
+        .onRestoreCompleted({ _ in
+            callbackOrder.append("onRestoreCompleted")
+        })
+        .onRestoreFailure({ _ in
+            callbackOrder.append("onRestoreFailure")
+        })
+        .addToHierarchy()
+
+        Task {
+            _ = try await purchasHandler.purchase(package: Self.package)
+        }
+
+        expect(callbackOrder).toEventually(equal(["onPurchaseStarted", "performPurchase", "onPurchaseFailure"]))
+        expect(passedError).toEventually(matchError(TestError.error))
+    }
+
+    func testHandleExternalRestore() throws {
+        var completed = false
+        var customerInfo: CustomerInfo?
+        var callbackOrder = [String]()
+
+        let purchasHandler = Self.externalPurchaseHandler { _ in
+            return (userCancelled: true, error: nil)
+        } performRestore: {
+            callbackOrder.append("performRestore")
+            return (success: true, error: nil)
+        }
+
+        try PaywallView(
+            offering: Self.offering.withLocalImages,
+            customerInfo: TestData.customerInfo,
+            introEligibility: .producing(eligibility: .eligible),
+            purchaseHandler: purchasHandler
+        )
+        .onPurchaseStarted { _ in
+            callbackOrder.append("onPurchaseStarted")
+        }
+        .onPurchaseCompleted { _ in
+            callbackOrder.append("onPurchaseCompleted")
+        }
+        .onPurchaseCancelled {
+            callbackOrder.append("onPurchaseCancelled")
+        }
+        .onPurchaseFailure { _ in
+            callbackOrder.append("onPurchaseFailure")
+        }
+        .onRestoreStarted({
+            callbackOrder.append("onRestoreStarted")
+        })
+        .onRestoreCompleted({ info in
+            callbackOrder.append("onRestoreCompleted")
+            customerInfo = info
+        })
+        .onRestoreFailure({ _ in
+            callbackOrder.append("onRestoreFailure")
+        })
+        .addToHierarchy()
+
+        Task {
+            _ = try await purchasHandler.restorePurchases()
+            completed = true
+        }
+
+        expect(completed).toEventually(beTrue())
+        expect(customerInfo).toEventually(be(TestData.customerInfo))
```

Comment:
also unrelated note: once the  first `toEventually` goes through, the rest can be regular expectations without polling, since Nimble will stop at the line until the first `toEventually` evaluates to `true`, and only then continue on. 

this means slightly faster tests, but also helps us ensure that we only expect stuff to be async once, not more than once

Reply:
Great, done!